src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+/// client), we ignore the input fds and open new fds directly to the tty path, which happens to
+/// be addressable as a file in OSX and Linux. In that case, there is no middle-man, and we
+/// ignore the fds opened by the nailgun server for data sent via the protocol.
+///
+#[derive(Clone)]
+struct RawFdNail {
+  executor: Executor,
+  runner: Arc<dyn RawFdRunner>,
+}
+
+impl Nail for RawFdNail {
+  fn spawn(
+    &self,
+    cmd: execution::Command,
+    input_stream: mpsc::Receiver<ChildInput>,
+  ) -> Result<Child, io::Error> {
+    let env = cmd.env.iter().cloned().collect::<HashMap<_, _>>();
+
+    // Handle stdin.
+    let (stdin_handle, stdin_sink) = Self::input(Self::ttypath_from_env(&env, 0))?;
+    let accepts_stdin = if let Some(mut stdin_sink) = stdin_sink {
+      let mut input_stream = input_stream.filter_map(|child_input| {
+        Box::pin(async move {
+          match child_input {
+            ChildInput::Stdin(bytes) => Some(Ok(bytes)),
+            ChildInput::StdinEOF => None,
+          }
+        })
+      });
+      let _join = self.executor.spawn(async move {
+        stdin_sink.send_all(&mut input_stream).map(|_| ()).await;
+      });
+      true
+    } else {
+      // Stdin will be handled directly by the TTY.
+      false
+    };
+
+    // And stdout/stderr.
+    let (stdout_stream, stdout_handle) = Self::output(Self::ttypath_from_env(&env, 1))?;
+    let (stderr_stream, stderr_handle) = Self::output(Self::ttypath_from_env(&env, 2))?;
+
+    // Spawn the underlying function as a blocking task, and capture its exit code to append to the
+    // output stream.
+    let nail = self.clone();
+    let exit_code_future = self.executor.spawn_blocking(move || {
+      // NB: This closure captures the stdio handles, and will drop/close them when it completes.
+      (nail.runner)(RawFdExecution {
+        cmd,
+        stdin_fd: stdin_handle.as_raw_fd(),
+        stdout_fd: stdout_handle.as_raw_fd(),
+        stderr_fd: stderr_handle.as_raw_fd(),
+      })
+    });
+
+    // Fully consume the stdout/stderr streams before waiting on the exit stream.
+    let stdout_stream = stdout_stream.map_ok(ChildOutput::Stdout);
+    let stderr_stream = stderr_stream.map_ok(ChildOutput::Stderr);
+    let exit_stream = exit_code_future
+      .into_stream()
+      .map(|exit_code| Ok(ChildOutput::Exit(exit_code)));
+    let output_stream = stream::select(stdout_stream, stderr_stream)
+      .chain(exit_stream)
+      .map(|res| match res {
+        Ok(o) => o,
+        Err(e) => {
+          error!("IO error interacting with the runner: {:?}", e);
+          ChildOutput::Exit(ExitCode(-1))
+        }
+      })
+      .boxed();
+
+    Ok(Child {
+      output_stream,
+      accepts_stdin,
+    })
+  }
+}
+
+impl RawFdNail {
+  ///
+  /// Returns a tuple of a readable file handle and an optional sink for nails to send stdin to.
+  ///
+  /// In the case of a TTY, the file handle will point directly to the TTY, and no stdin data will
```

Comment:
As above, might be good to put a TODO for an explicit enum here instead of `Option<_>`. That's an opinion, but I find it useful to break out enums to move fast and not break things, personally.